From de7ad644d65b4c569e98c1af57ef377c23e1f409 Mon Sep 17 00:00:00 2001
From: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
Date: Wed, 21 May 2014 15:19:37 +0200
Subject: [PATCH 5/5] omap3: overo: Add usb host support

Series-to: u-boot
Series-cc: Ash Charles <ash@gumstix.com>
Series-cc: Steve Sakoman <sakoman@gmail.com>
Series-cc: Tom Rini <trini@ti.com>
Cover-letter
omap3: overo: Add foreign expansion board support

At the moment the boot loader use a common configuration for all Gumstix
Overo expansion boards. This also includes configuration of all inputs
and outputs even if they aren't used by the boot loader. Because the
Overo module could be used on foreign expansion boards with different
configurations this could lead to problems. Split the configuration into
a Overo module and expansion board part. The expansion board
configuration is skipped if a foreign expansion board is detected. This
enables a foreign expansion board to load a specific device tree by
saving its name in the EEPROM. Additionally fix some errors in the
environment configuration and add USB support.

The series is only tested on a foreign expansion boards. It would be nice
if somebody could test it on a expansion board with network support.
END

Signed-off-by: Stefan Herbrechtsmeier <stefan@herbrechtsmeier.net>
---
 board/overo/overo.c           | 34 ++++++++++++++++++++++++++++++++++
 include/configs/omap3_overo.h |  8 ++++++++
 2 files changed, 42 insertions(+)

diff --git a/board/overo/overo.c b/board/overo/overo.c
index 488246b..66146ee 100644
--- a/board/overo/overo.c
+++ b/board/overo/overo.c
@@ -25,6 +25,11 @@
 #include <asm/mach-types.h>
 #include "overo.h"
 
+#ifdef CONFIG_USB_EHCI
+#include <usb.h>
+#include <asm/ehci-omap.h>
+#endif
+
 DECLARE_GLOBAL_DATA_PTR;
 
 #define TWL4030_I2C_BUS			0
@@ -474,3 +479,32 @@ int board_mmc_init(bd_t *bis)
 	return omap_mmc_init(0, 0, 0, -1, -1);
 }
 #endif
+
+#if defined(CONFIG_USB_EHCI) &&  !defined(CONFIG_SPL_BUILD)
+static struct omap_usbhs_board_data usbhs_bdata = {
+	.port_mode[0] = OMAP_USBHS_PORT_MODE_UNUSED,
+	.port_mode[1] = OMAP_EHCI_PORT_MODE_PHY,
+	.port_mode[2] = OMAP_USBHS_PORT_MODE_UNUSED
+};
+
+#define GUMSTIX_GPIO_USBH_CPEN		168
+int ehci_hcd_init(int index, enum usb_init_type init,
+		  struct ehci_hccr **hccr, struct ehci_hcor **hcor)
+{
+	/* Enable USB power */
+	if (!gpio_request(GUMSTIX_GPIO_USBH_CPEN, "usbh_cpen"))
+		gpio_direction_output(GUMSTIX_GPIO_USBH_CPEN, 1);
+
+	return omap_ehci_hcd_init(index, &usbhs_bdata, hccr, hcor);
+}
+
+int ehci_hcd_stop(void)
+{
+	/* Disable USB power */
+	gpio_set_value(GUMSTIX_GPIO_USBH_CPEN, 0);
+	gpio_free(GUMSTIX_GPIO_USBH_CPEN);
+
+	return omap_ehci_hcd_stop();
+}
+
+#endif /* CONFIG_USB_EHCI */
diff --git a/include/configs/omap3_overo.h b/include/configs/omap3_overo.h
index d042eea..c58636a 100644
--- a/include/configs/omap3_overo.h
+++ b/include/configs/omap3_overo.h
@@ -35,6 +35,13 @@
 /* TWL4030 LED */
 #define CONFIG_TWL4030_LED
 
+/* USB EHCI */
+#define CONFIG_USB_EHCI
+#define CONFIG_USB_EHCI_OMAP
+#define CONFIG_USB_STORAGE
+#define CONFIG_OMAP_EHCI_PHY1_RESET_GPIO	183
+#define CONFIG_SYS_USB_EHCI_MAX_ROOT_PORTS	3
+
 /* Initialize GPIOs by default */
 #define CONFIG_OMAP3_GPIO_2	/* GPIO32..63 is in GPIO Bank 2 */
 #define CONFIG_OMAP3_GPIO_3	/* GPIO64..95 is in GPIO Bank 3 */
@@ -44,6 +51,7 @@
 
 /* commands to include */
 #define CONFIG_CMD_CACHE
+#define CONFIG_CMD_USB
 #undef CONFIG_CMD_FPGA		/* FPGA configuration Support	*/
 #undef CONFIG_CMD_IMI		/* iminfo			*/
 #undef CONFIG_CMD_NFS		/* NFS support			*/
-- 
2.1.0

